File: algorithms/Longest Turbulent Subarray/Longest Turbulent Subarray.java
class Solution {
    public int maxTurbulenceSize(int[] arr) {
        if(arr.length == 1) {
            return 1;
        } 
        int l = 0, r = 1;
        int diff = arr[l] - arr[r];
        int max;
        if(diff == 0) {
            l = 1;
            r = 1;
            max = 1;
        } else {
            l = 0;
            r = 1;
            max = 2;
        }
        for(int i = 1; r < arr.length-1; i++) {
            int nextdiff = arr[i] - arr[i+1];
            if(diff < 0) {
                if(nextdiff > 0) {
                    r++;
                } else if(nextdiff == 0) {
                    l = i+1;
                    r = i+1;
                } else {
                    l = i;
                    r = i+1;
                }
            } else {
                if(nextdiff < 0) {
                    r++;
                } else if(nextdiff == 0) {
                    l = i+1;
                    r = i+1;
                } else {
                    l = i;
                    r = i+1;
                }
            }
            diff = nextdiff;
            max = Math.max(max, r-l+1);
        }
        return max;
    }
}

File: scripts/algorithms/L/Longest Turbulent Subarray/Longest Turbulent Subarray.cpp
class Solution {
public:
    int maxTurbulenceSize(vector<int>& arr) {
        vector<int> table1(arr.size(), 0);
        vector<int> table2(arr.size(), 0);
        table1[0] = 1;
        table2[0] = 1;
        int max_len = 1;
        for (int i=1; i<arr.size(); ++i) {
            table1[i] = 1;
            table2[i] = 1;
            if (arr[i] < arr[i - 1] && (i & 1) == 0) {
                table1[i] = table1[i - 1] + 1;
            } else if (arr[i] > arr[i - 1] && (i & 1) == 1) {
                table1[i] = table1[i - 1] + 1;
            }
            if (arr[i] > arr[i - 1] && (i & 1) == 0) {
                table2[i] = table2[i - 1] + 1;
            } else if (arr[i] < arr[i - 1] && (i & 1) == 1) {
                table2[i] = table2[i - 1] + 1;
            }
            max_len = max(max_len, table1[i]);
            max_len = max(max_len, table2[i]);
        }
        return max_len;
    }
};
